onnxruntime/core/providers/cpu/nn/qlinearconv.cc

```diff
@@ -415,134 +422,144 @@ Status QLinearConv::Compute(OpKernelContext* context) const {
       auto* worker_gemm_output = gemm_output + output_start * M;
       auto* worker_requantize_output = output_data + output_start * M;
 
-      for (int64_t group_id = 0; group_id < group_count; ++group_id) {
-        // Prepare the im2col transformation or use the input buffer directly for
-        // pointwise convolutions.
-        const uint8_t* worker_gemm_input;
-        if (col_buffer) {
-          auto* worker_col_buffer = static_cast<uint8_t*>(col_buffer.get()) + output_start * kernel_dim;
-          if (kernel_rank == 2) {
-            math::Im2col<uint8_t, StorageOrder::NHWC>()(
-                input_data + group_id * group_input_channels,
-                group_input_channels,
-                C,
-                input_shape[0],
-                input_shape[1],
-                kernel_shape[0],
-                kernel_shape[1],
-                dilations[0],
-                dilations[1],
-                pads[0],
-                pads[1],
-                strides[0],
-                strides[1],
-                output_shape[1],
-                output_start,
-                output_count,
-                worker_col_buffer,
-                X_zero_point_value);
-          } else if (kernel_rank == 1) {
-            math::Im2col<uint8_t, StorageOrder::NHWC>()(
-                input_data + group_id * group_input_channels,
-                group_input_channels,
-                C,
-                1,
-                input_shape[0],
-                1,
-                kernel_shape[0],
-                1,
-                dilations[0],
-                0,
-                pads[0],
-                1,
-                strides[0],
-                output_shape[0],
-                output_start,
-                output_count,
-                worker_col_buffer,
-                X_zero_point_value);
+      if (is_depthwise_conv) {
+        auto* worker_col_buffer = static_cast<uint8_t const**>(col_buffer.get()) + output_start * kernel_size;
+        math::Im2col<uint8_t, StorageOrder::NHWC>()(
+            input_data,
+            C,
+            input_shape.GetDims().data(),
+            output_shape.GetDims().data(),
+            kernel_shape.data(),
+            strides.data(),
+            dilations.data(),
+            pads.data(),
+            static_cast<ptrdiff_t>(kernel_rank),
+            output_start,
+            output_count,
+            worker_col_buffer,
+            padding_data.data());
```

Comment:
The input data is continuous. Let's say the size of one image is NSize. Will it be faster to just add NSize on the indirection buffer for each iteration?